src/components/Table/utils/RowState.ts

```diff
@@ -0,0 +1,136 @@
+   *
+   * I.e. a row might have `inferSelectedState: false`, so is showing unchecked, but the header
+   * wants to show partial-ness whenever any given child is selected.
+   */
+  get selectedStateForHeader(): SelectedState {
+    if (this.children) {
+      const children = this.visibleChildren;
+      const allChecked = children.every((child) => child.selectedStateForHeader === "checked");
+      const allUnchecked = children.every((child) => child.selectedStateForHeader === "unchecked");
+      // For the header purposes, if this is a selectable-row (i.e. not inferSelectedState) make sure
+      // both the row itself & all children are "all checked" or "not all checked", otherwise consider
+      // ourselves partially selected.
+      if ((allUnchecked || children.length === 0) && (this.inferSelectedState || !this.selected)) {
+        return "unchecked";
+      } else if (allChecked && (this.inferSelectedState || this.selected)) {
+        return "checked";
+      } else {
+        return "partial";
+      }
+    }
+    return this.selected ? "checked" : "unchecked";
+  }
+
+  /**
+   * Called to explicitly select/unselect this row.
+   *
+   * This could be either because the user clicked directly on us, or because we're a visible
+   * child of a selected parent row.
+   */
+  select(selected: boolean): void {
+    if (this.row.selectable === false) return;
+    this.selected = selected;
+    // We don't check inferSelectedState here, b/c even if the parent is considered selectable
+    // on its own, we still push down selected-ness to our visible children.
+    if (this.children) {
+      for (const child of this.visibleChildren) {
+        child.select(selected);
+      }
+    }
+  }
+
+  /** Whether this is a selected-but-filtered-out row that we should hoist to the top. */
+  get isKept(): boolean {
+    // this row is "kept" if it is selected, and:
+    // - it is not matched (hidden by filter) (being hidden by collapse is okay)
+    // - or it has (probably) been server-side filtered
+    return (
+      this.selected &&
+      // Headers, totals, etc., do not need keeping
+      !reservedRowKinds.includes(this.row.kind) &&
+      // Parents don't need keeping, unless they're actually real rows
+      !(this.children && this.inferSelectedState) &&
+      (!this.isMatched || this.wasRemoved)
+    );
+  }
+
+  private get inferSelectedState(): boolean {
+    return this.row.inferSelectedState !== false;
+  }
+
+  private get visibleChildren(): RowState[] {
+    // The keptGroup should treat all of its children as visible, as this makes select/unselect all work.
+    if (this.row.kind === KEPT_GROUP) return this.children ?? [];
+    return this.children?.filter((c) => c.isMatched === true) ?? [];
+  }
+
+  /** Pretty toString. */
+  [Symbol.for("nodejs.util.inspect.custom")](): string {
+    return `RowState ${this.row.kind}-${this.row.id}`;
+  }
```

Comment:
**question:** What do we use this for?

Reply:
When the `RowState` ends up in `console.log("something ${rs} something")` then Node will call this ... actually, no, I think it's _not_ string interpolation, like `console.log("this", rs)` goes through this symbol... It was useful for some debugging I was doing, but I should clarify in my head + document exactly where it's used.